simpledb/plan/sort_plan.go

```diff
@@ -0,0 +1,229 @@
+package plan
+
+import (
+	"fmt"
+	"simpledb/query"
+	"simpledb/record"
+	"simpledb/tx"
+)
+
+var _ Plan = (*SortPlan)(nil)
+
+type SortPlan struct {
+	plan   Plan
+	tx     *tx.Transaction
+	schema *record.Schema
+	comp   *query.RecordComparator
+}
+
+func NewSortPlan(tx *tx.Transaction, plan Plan, sortFields []string) (*SortPlan, error) {
+	return &SortPlan{
+		plan:   plan,
+		tx:     tx,
+		schema: plan.Schema(),
+		comp:   query.NewRecordComparator(sortFields),
+	}, nil
+}
+
+func (sp *SortPlan) Open() (query.Scan, error) {
+	src, err := sp.plan.Open()
+	if err != nil {
+		return nil, fmt.Errorf("sp.plan.Open: %w", err)
+	}
+	runs, err := sp.splitIntoRuns(src)
+	if err != nil {
+		return nil, fmt.Errorf("sp.splitIntoRuns: %w", err)
+	}
+	src.Close()
+	for len(runs) > 2 {
+		runs, err = sp.doAMergeIteration(runs)
+		if err != nil {
+			return nil, fmt.Errorf("sp.doAMergeIteration: %w", err)
+		}
+	}
+
+	sc, err := query.NewSortScan(runs, sp.comp)
+	if err != nil {
+		return nil, fmt.Errorf("query.NewSortScan: %w", err)
+	}
+	return sc, nil
+}
+
+func (sp *SortPlan) BlocksAccessed() int {
+	mp := NewMaterializePlan(sp.tx, sp.plan)
+	return mp.BlocksAccessed()
+}
+
+func (sp *SortPlan) RecordsOutput() int {
+	return sp.plan.RecordsOutput()
+}
+
+func (sp *SortPlan) DistinctValues(fieldName string) int {
+	return sp.plan.DistinctValues(fieldName)
+}
+
+func (sp *SortPlan) Schema() *record.Schema {
+	return sp.schema
+}
+
+func (sp *SortPlan) splitIntoRuns(src query.Scan) ([]*query.TempTable, error) {
+	temps := make([]*query.TempTable, 0)
+	err := src.BeforeFirst()
+	if err != nil {
+		return nil, fmt.Errorf("src.BeforeFirst: %w", err)
+	}
+
+	next, err := src.Next()
+	if err != nil {
+		return nil, err
+	}
+	if !next {
+		return temps, nil
+	}
+
+	currentTemp := query.NewTempTable(sp.tx, sp.schema)
+	temps = append(temps, currentTemp)
+	currentScan, err := currentTemp.Open()
+	if err != nil {
+		return nil, err
+	}
+
+	// src < currentScan になるまでコピーし続けることで、
+	// 昇順にソートされたtemp table（ラン）が作成される。
+	// 順序が崩れたら新しいtemp tableにコピーしていく。
```

Comment:
[imo] コメントの説明がむずかしいところは例を書くの好き勢です

```suggestion
	// 順序が崩れたら新しいtemp tableにコピーしていく。
	// ex. [2, 6, 20, 4, 1, 16, 19, 3, 18] => [2, 6, 10], [4], [1, 16, 19], [3, 18]
```